x-pack/plugins/enterprise_search/server/lib/ml/fetch_ml_models.ts

```diff
@@ -0,0 +1,236 @@
+  targetAllocationCount: 0,
+  threadsPerAllocation: 0,
+  isPlaceholder: false,
+  hasStats: false,
+};
+
+const ELSER_V2_MODEL_PLACEHOLDER: MlModel = {
+  ...BASE_MODEL,
+  modelId: ELSER_MODEL_ID,
+  type: SUPPORTED_PYTORCH_TASKS.TEXT_EXPANSION,
+  title: 'Elastic Learned Sparse EncodeR (ELSER)',
+  description: i18n.translate('xpack.enterpriseSearch.modelCard.elserPlaceholder.description', {
+    defaultMessage:
+      'ELSER is designed to efficiently use context in natural language queries with better results than BM25 alone.',
+  }),
+  license: 'Elastic',
+  isPlaceholder: true,
+};
+
+const E5_MODEL_PLACEHOLDER: MlModel = {
+  ...BASE_MODEL,
+  modelId: E5_MODEL_ID,
+  type: SUPPORTED_PYTORCH_TASKS.TEXT_EMBEDDING,
+  title: 'E5 Multilingual Embedding',
+  description: i18n.translate('xpack.enterpriseSearch.modelCard.e5Placeholder.description', {
+    defaultMessage: 'Multilingual dense vector embedding generator.',
+  }),
+  license: 'MIT',
+  modelDetailsPageUrl: 'https://huggingface.co/intfloat/multilingual-e5-small',
+  isPlaceholder: true,
+};
+
+/**
+ * Fetches and enriches trained model information and deployment status. Pins promoted models (ELSER, E5) to the top. If a promoted model doesn't exist, a placeholder will be used.
+ *
+ * @param trainedModelsProvider Trained ML models provider
+ * @returns List of models
+ */
+export const fetchMlModels = async (
+  trainedModelsProvider: MlTrainedModels | undefined
+): Promise<MlModel[]> => {
+  if (!trainedModelsProvider) {
+    throw new Error('Machine Learning is not enabled');
+  }
+
+  // This array will contain all models, let's add placeholders first
+  const models: MlModel[] = [ELSER_V2_MODEL_PLACEHOLDER, E5_MODEL_PLACEHOLDER];
+
+  // Fetch all models and their deployment stats using the ML client
+  const modelsResponse = await trainedModelsProvider.getTrainedModels({});
+  if (modelsResponse.count === 0) {
+    return models;
+  }
+  const modelsStatsResponse = await trainedModelsProvider.getTrainedModelsStats({});
+
+  modelsResponse.trained_model_configs
+    // Filter unsupported models
+    .filter((modelConfig) => isSupportedModel(modelConfig))
+    // Get corresponding model stats and compose full model object
+    .map((modelConfig) =>
+      getModel(
+        modelConfig,
+        modelsStatsResponse.trained_model_stats.find((m) => m.model_id === modelConfig.model_id)
+      )
+    )
+    // Merge models with placeholders
+    // (Note: properties from the placeholder that are undefined in the model are preserved)
+    .forEach((model) => mergeModel(model, models));
+
+  // Undeployed placeholder models might be in the Downloading phase; let's evaluate this with a call
+  // We must do this one by one because the API doesn't support fetching multiple models with include=definition_status
+  for (const model of models) {
+    if (model.isPromoted && !model.isPlaceholder && !model.hasStats) {
+      await enrichModelWithDownloadStatus(model, trainedModelsProvider);
```

Comment:
This is sub-optimal (blocking code in a for loop). I'm thinking we should use `Promise.all()` here to run the fetches in parallel. Thoughts?

Reply:
Hm, so the explicit `await` won't force the engine to wait for each `enrichModelWithDownloadStatus()`'s result in sequence? The concern is that some enrichment API calls can take up to a second, so it's best to execute them in parallel. Are you saying something like this could lead to issues?

```js
enrichedModels = await Promise.all([enrich(model1), enrich(model1)...])
```

Reply:
Explicit `await` will force the "call to this function" to wait, but it won't block server to be running. As you said it will wait one to be done and then do the other and then do the other ...


So if you have 500 models, `Promise.all` would run them all at the same time. If it is 10 probably it is fine. As you said it can take up to a few seconds with current implementation. 

Probably `p-limit` is the best option if you are worried about the performance of the call and also to make sure we are not going to fail with large amount of calls.